### PR TITLE
datanode bootstrap implementation

### PR DESCRIPTION
**Pull request by jshencode**
datanode bootstrap implementation
Open() -> Bootstrap() -> Serve()
more tests on datanode are being added

major changes:
memstore/bootstrap.go
datanode/bootstrap_manager.go
datanode/datanode.go

following up:
1. datanode cmd and main
2. peer fetch failure handling and retry

### CODE REVIEW

File: memstore/memstore.go [hunk 1]
```diff
@@ -252,3 +258,92 @@ func (m *memStoreImpl) TryEvictBatchColumn(table string, shardID int, batchID in
...
+	// Destruct.
+	if shard != nil {
+		shard.Destruct()
+	}
+}
+
+// preloadAllFactTables preloads recent days data for all columns of all table shards into memory.
+// The number of preloading days is defined at each column level. This call will happen at
+// shard initialization stage.
+func (m *memStoreImpl) preloadAllFactTables() {
```
Comment (jshencode) [inline memstore/memstore.go]: not new code, move preloadAll() within host memory manager.Start() out to the memstore level

File: datanode/datanode.go [hunk 2]
```diff
@@ -15,303 +15,443 @@
...
+
+	for _, shardID := range removing {
+		for _, table := range tables {
+			d.memStore.RemoveTableShard(table, int(shardID))
+		}
+	}
+
+	for _, shard := range adding {
+		for _, table := range tables {
+			d.memStore.AddTableShard(table, int(shard.ID()), shard.State() == m3Shard.Initializing)
```
Comment (voyager-dw) [inline datanode/datanode.go]: AddTableShard only after bootstrap, while shard change triggers bootstrap/recovery process?

File: datanode/datanode.go [hunk 3]
```diff
@@ -15,303 +15,443 @@
...
-func (d *dataNode) Options() Options {
-	return d.opts
+	return d, nil
 }
 
-// ShardSet returns the set of shards currently associated with this datanode.
-func (d *dataNode) ShardSet() (shard.ShardSet, error) {
-	t, err := d.namespace.Topology()
+// Open data node for serving
+func (d *dataNode) Open() error {
```
Comment (voyager-dw) [inline datanode/datanode.go]: Start()?

File: datanode/bootstrap.go [hunk 4]
```diff
@@ -129,12 +128,86 @@ func (m *bootstrapManagerImpl) Report() {
...
+	workers.Init()
+
+	var (
+		multiErr = xerrors.NewMultiError()
+		mutex    sync.Mutex
+		wg       sync.WaitGroup
+	)
+
+	topoStateSnapshot := newInitialTopologyState(m.topo)
+	for _, shard := range shardSet.All() {
```
[the remark this answers is not in the document]
Comment (jshencode) [reply datanode/bootstrap.go]: this is already the shardSet owned by this data node

File: common/config.go [hunk 5]
```diff
@@ -58,6 +58,11 @@ type GatewayConfig struct {
 	Controller *ControllerConfig `yaml:"controller,omitempty"`
 }
 
+type InstanceConfig struct {
+	ID string `yaml:"id"`
+	Zone string `yaml:"zone"`
```
[the remark this answers is not in the document]
Comment (jshencode) [reply common/config.go]: will add namespace later

File: memstore/job_manager.go [hunk 6]
```diff
@@ -62,7 +62,7 @@ func (m *archiveJobManager) generateJobs() []Job {
 	for tableName, shardMap := range m.memStore.TableShards {
```
[the remark this answers is not in the document]
Comment (jshencode) [reply memstore/job_manager.go]: then you need to place not available table shards in another place other then memStore, and thus another set of logic for managing them and move around table shards, i don't think that is necessary. It is actually easier to just add a call for get the available table shards within memstore

**Pull request merged**